fedimint-api/src/db/mod.rs

```diff
@@ -79,14 +84,76 @@ impl Database {
     pub fn new(db: impl IDatabase + 'static, module_decoders: ModuleDecoderRegistry) -> Self {
         let inner = DatabaseInner {
             db,
+            notifications: Notifications::new(),
             module_decoders,
         };
 
         Self(Arc::new(inner))
     }
 
     pub async fn begin_transaction(&self) -> DatabaseTransaction {
-        DatabaseTransaction::new(self.0.db.begin_transaction().await, &self.0.module_decoders)
+        DatabaseTransaction::new(
+            self.0.db.begin_transaction().await,
+            &self.0.module_decoders,
+            &self.0.notifications,
+        )
+    }
+
+    /// Waits for key to be notified.
+    ///
+    /// Calls the `f` on every notifications. Return the value when `f` returns a `Some(T)`
+    pub async fn wait_key_check<K, T>(
+        &self,
+        key: &K,
+        f: impl Fn(Option<K::Value>) -> Option<T>,
+    ) -> Result<T>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        // not compile time yet :/
+        assert!(
+            <K as DatabaseKey>::NOTIFY,
+            "wait_key called with key({}) without NOTIFY flag",
+            std::any::type_name::<K>()
+        );
+        let key_bytes = key.to_bytes();
+        loop {
+            // register for notification
+            let notify = self.0.notifications.wait(&key_bytes);
+
+            // check for value in db
+            let mut tx = self.0.db.begin_transaction().await;
+
+            let maybe_value_bytes = tx
+                .raw_get_bytes(&key_bytes)
+                .await?
+                .map(|value_bytes| {
+                    trace!(
+                        "get_value: Decoding {} from bytes {:?}",
+                        std::any::type_name::<K::Value>(),
+                        value_bytes
+                    );
+                    K::Value::from_bytes(&value_bytes, &self.0.module_decoders)
+                })
+                .transpose()?;
+
+            if let Some(value) = f(maybe_value_bytes) {
+                break Ok(value);
```

Comment:
I wonder if we need to return the transaction along with the value. I'd expect in some cases the user wants to make sure the nothing is going to mess withe db in between when they checked.

Reply:
_prays to the lifetime Gods_

Reply:
done, but only in `wait_key_check`